File: sdk/python/youtube_subtitle_api/webhook.py
"""
Webhook utilities for the YouTube Subtitle API SDK.

This module provides functions for verifying webhook signatures
and parsing webhook payloads.
"""

from __future__ import annotations

import hashlib
import hmac
import json
from typing import Any, Optional

from .models import WebhookEvent


def verify_signature(
    payload: bytes | str,
    signature: str,
    secret: str,
    timestamp: Optional[str] = None,
) -> bool:
    """
    Verify the HMAC signature of a webhook payload.

    The API signs webhooks using HMAC-SHA256. The signature is computed
    from the payload JSON plus an optional timestamp for freshness.

    Args:
        payload: The raw webhook payload (bytes or string)
        signature: The X-Webhook-Signature header value (e.g., "sha256=abc...")
        secret: Your webhook secret key
        timestamp: Optional X-Webhook-Timestamp header value

    Returns:
        True if signature is valid, False otherwise

    Example:
        >>> from fastapi import Request
        >>>
        >>> @app.post("/webhook")
        >>> async def handle_webhook(request: Request):
        ...     payload = await request.body()
        ...     sig = request.headers.get("X-Webhook-Signature", "")
        ...     ts = request.headers.get("X-Webhook-Timestamp", "")
        ...
        ...     if not verify_signature(payload, sig, WEBHOOK_SECRET, ts):
        ...         raise HTTPException(status_code=401, detail="Invalid signature")
        ...
        ...     event = parse_webhook(payload)
        ...     return {"status": "received"}
    """
    # Extract signature hash (remove "sha256=" prefix if present)
    if signature.startswith("sha256="):
        signature_hash = signature[7:]
    else:
        signature_hash = signature

    # Ensure payload is bytes for consistent hashing
    if isinstance(payload, str):
        payload_bytes = payload.encode("utf-8")
    else:
        payload_bytes = payload

    # Create the message to sign
    # The API signs: payload_json + "." + timestamp
    if timestamp:
        message = f"{payload_bytes.decode('utf-8')}.{timestamp}"
    else:
        message = payload_bytes.decode("utf-8")

    # Compute expected signature
    expected_hash = hmac.new(
        secret.encode("utf-8"),
        message.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()

    # Use constant-time comparison to prevent timing attacks
    return hmac.compare_digest(signature_hash, expected_hash)


def parse_webhook(payload: bytes | str | dict[str, Any]) -> WebhookEvent:
    """
    Parse a webhook payload into a WebhookEvent object.

    Args:
        payload: The webhook payload (bytes, string, or dict)

    Returns:
        WebhookEvent object with the parsed data

    Raises:
        ValueError: If payload is invalid

    Example:
        >>> @app.post("/webhook")
        >>> async def handle_webhook(request: Request):
        ...     payload = await request.body()
        ...     event = parse_webhook(payload)
        ...
        ...     if event.is_success:
        ...         subtitle = event.subtitle
        ...         print(f"Got {len(subtitle.subtitles)} subtitle items")
        ...     else:
        ...         print(f"Job failed: {event.error}")
    """
    # Parse JSON if needed
    if isinstance(payload, dict):
        data = payload
    elif isinstance(payload, bytes):
        data = json.loads(payload.decode("utf-8"))
    elif isinstance(payload, str):
        data = json.loads(payload)
    else:
        raise ValueError(f"Invalid payload type: {type(payload)}")

    return WebhookEvent.from_dict(data)


def verify_and_parse_webhook(
    payload: bytes | str,
    signature: str,
    secret: str,
    timestamp: Optional[str] = None,
) -> WebhookEvent:
    """
    Verify signature and parse webhook payload in one step.

    This is a convenience function that combines verify_signature()
    and parse_webhook().

    Args:
        payload: The raw webhook payload
        signature: The X-Webhook-Signature header value
        secret: Your webhook secret key
        timestamp: Optional X-Webhook-Timestamp header value

    Returns:
        WebhookEvent object with the parsed data

    Raises:
        ValueError: If signature is invalid or payload cannot be parsed

    Example:
        >>> @app.post("/webhook")
        >>> async def handle_webhook(request: Request):
        ...     payload = await request.body()
        ...     sig = request.headers.get("X-Webhook-Signature", "")
        ...     ts = request.headers.get("X-Webhook-Timestamp", "")
        ...
        ...     try:
        ...         event = verify_and_parse_webhook(payload, sig, WEBHOOK_SECRET, ts)
        ...     except ValueError:
        ...         raise HTTPException(status_code=401, detail="Invalid signature")
        ...
        ...     return process_event(event)
    """
    if not verify_signature(payload, signature, secret, timestamp):
        raise ValueError("Invalid webhook signature")

    return parse_webhook(payload)


def generate_signature(
    payload: dict[str, Any] | str,
    secret: str,
    timestamp: Optional[str] = None,
) -> str:
    """
    Generate an HMAC signature for a webhook payload.

    This is useful for testing webhook handlers or for making
    signed webhook requests to other services.

    Args:
        payload: The payload to sign (dict or string)
        secret: The webhook secret key
        timestamp: Optional timestamp to include in signature

    Returns:
        The signature hash (without "sha256=" prefix)

    Example:
        >>> payload = {"event": "job.completed", "job_id": "123"}
        >>> sig = generate_signature(payload, "my_secret")
        >>> headers = {"X-Webhook-Signature": f"sha256={sig}"}
    """
    if isinstance(payload, dict):
        payload_json = json.dumps(payload, sort_keys=True, separators=(",", ":"))
    else:
        payload_json = payload

    if timestamp:
        message = f"{payload_json}.{timestamp}"
    else:
        message = payload_json

    signature_hash = hmac.new(
        secret.encode("utf-8"),
        message.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()

    return signature_hash


class WebhookVerifier:
    """
    A reusable webhook verifier with pre-configured secret.

    Example:
        >>> verifier = WebhookVerifier(secret="my_webhook_secret")
        >>>
        >>> @app.post("/webhook")
        >>> async def handle_webhook(request: Request):
        ...     payload = await request.body()
        ...     sig = request.headers.get("X-Webhook-Signature", "")
        ...     ts = request.headers.get("X-Webhook-Timestamp", "")
        ...
        ...     event = verifier.verify_and_parse(payload, sig, ts)
        ...     return {"status": "received"}
    """

    def __init__(self, secret: str, require_timestamp: bool = False):
        """
        Initialize the verifier.

        Args:
            secret: Your webhook secret key
            require_timestamp: Whether to require timestamp in signatures
        """
        self.secret = secret
        self.require_timestamp = require_timestamp

    def verify(
        self,
        payload: bytes | str,
        signature: str,
        timestamp: Optional[str] = None,
    ) -> bool:
        """
        Verify a webhook signature.

        Args:
            payload: The raw webhook payload
            signature: The X-Webhook-Signature header value
            timestamp: Optional X-Webhook-Timestamp header value

        Returns:
            True if signature is valid
        """
        if self.require_timestamp and not timestamp:
            return False

        return verify_signature(payload, signature, self.secret, timestamp)

    def parse(self, payload: bytes | str | dict[str, Any]) -> WebhookEvent:
        """
        Parse a webhook payload.

        Args:
            payload: The webhook payload

        Returns:
            WebhookEvent object
        """
        return parse_webhook(payload)

    def verify_and_parse(
        self,
        payload: bytes | str,
        signature: str,
        timestamp: Optional[str] = None,
    ) -> WebhookEvent:
        """
        Verify signature and parse webhook payload.

        Args:
            payload: The raw webhook payload
            signature: The X-Webhook-Signature header value
            timestamp: Optional X-Webhook-Timestamp header value

        Returns:
            WebhookEvent object

        Raises:
            ValueError: If signature is invalid
        """
        if not self.verify(payload, signature, timestamp):
            raise ValueError("Invalid webhook signature")

        return self.parse(payload)

File: sdk/typescript/src/webhook.ts
/**
 * Webhook utilities for YouTube Subtitle API SDK
 */

import crypto from "node:crypto";
import type {
  WebhookPayload,
  Subtitle,
  BatchExtractResponse,
  JobStatus,
} from "./types.js";

/**
 * Verify webhook signature
 *
 * @param payload - The raw webhook payload string
 * @param signature - The signature from the X-Signature header
 * @param secret - Your webhook secret
 * @returns True if the signature is valid
 */
export function verifySignature(
  payload: string,
  signature: string,
  secret: string,
): boolean {
  if (!signature || !secret) {
    return false;
  }

  // Remove 'sha256=' prefix if present
  const signatureBytes = signature.replace(/^sha256=/i, "");

  // Compute HMAC
  const hmac = crypto.createHmac("sha256", secret);
  hmac.update(payload);
  const expectedSignature = hmac.digest("hex");

  // Constant-time comparison to prevent timing attacks
  return crypto.timingSafeEqual(
    Buffer.from(signatureBytes),
    Buffer.from(expectedSignature),
  );
}

/**
 * Parse and validate a webhook payload
 *
 * @param payload - The raw webhook payload (string or already parsed object)
 * @returns The parsed and validated webhook payload
 * @throws ValidationError if the payload is invalid
 */
export function parseWebhook(
  payload: string | Record<string, unknown>,
): WebhookPayload {
  let data: Record<string, unknown>;

  // Parse JSON if string
  if (typeof payload === "string") {
    try {
      data = JSON.parse(payload) as Record<string, unknown>;
    } catch {
      throw new Error("Invalid JSON payload");
    }
  } else {
    data = payload;
  }

  // Validate required fields
  if (!data.event || typeof data.event !== "string") {
    throw new Error("Missing or invalid 'event' field");
  }

  const validEvents = [
    "subtitle.completed",
    "subtitle.failed",
    "batch.completed",
  ] as const;

  if (!validEvents.includes(data.event as any)) {
    throw new Error(`Invalid event type: ${data.event}`);
  }

  if (!data.job_id || typeof data.job_id !== "string") {
    throw new Error("Missing or invalid 'job_id' field");
  }

  if (!data.timestamp) {
    throw new Error("Missing 'timestamp' field");
  }

  if (!data.data || typeof data.data !== "object") {
    throw new Error("Missing or invalid 'data' field");
  }

  return data as WebhookPayload;
}

/**
 * Webhook event handler type
 */
export type WebhookEventHandler = (
  payload: WebhookPayload,
) => void | Promise<void>;

/**
 * Webhook handler configuration
 */
export interface WebhookHandlerConfig {
  /** Webhook secret for signature verification */
  secret: string;
  /** Handler for subtitle completed events */
  onSubtitleCompleted?: (result: Subtitle) => void | Promise<void>;
  /** Handler for subtitle failed events */
  onSubtitleFailed?: (error: {
    job_id: string;
    error: string;
  }) => void | Promise<void>;
  /** Handler for batch completed events */
  onBatchCompleted?: (result: BatchExtractResponse) => void | Promise<void>;
  /** Optional handler for all events */
  onAny?: (payload: WebhookPayload) => void | Promise<void>;
  /** Optional logger */
  logger?: {
    info?: (message: string, ...args: unknown[]) => void;
    error?: (message: string, ...args: unknown[]) => void;
  };
}

/**
 * Create a webhook handler function for use with Express, Fastify, etc.
 *
 * @example
 * ```ts
 * import express from 'express';
 * import { createWebhookHandler } from '@youtube-subtitle-api/sdk/webhook';
 *
 * const app = express();
 * app.use(express.raw({ type: 'application/json' }));
 *
 * const handler = createWebhookHandler({
 *   secret: process.env.WEBHOOK_SECRET!,
 *   onSubtitleCompleted: (result) => {
 *     console.log('Subtitles extracted:', result.video_id);
 *   },
 * });
 *
 * app.post('/webhook', handler);
 * ```
 */
export function createWebhookHandler(
  config: WebhookHandlerConfig,
): (req: {
  body: Buffer;
  headers: Record<string, string>;
}) => Promise<Response> {
  const {
    secret,
    onSubtitleCompleted,
    onSubtitleFailed,
    onBatchCompleted,
    onAny,
    logger = console,
  } = config;

  return async (req): Promise<Response> => {
    try {
      const body = req.body.toString("utf-8");
      const signature = req.headers["x-signature"] || "";

      // Verify signature
      if (!verifySignature(body, signature, secret)) {
        logger.error?.("Invalid webhook signature");
        return new Response("Invalid signature", { status: 401 });
      }

      // Parse payload
      const payload = parseWebhook(body);

      logger.info?.(`Webhook received: ${payload.event}`, {
        job_id: payload.job_id,
      });

      // Call specific handlers
      switch (payload.event) {
        case "subtitle.completed":
          await onSubtitleCompleted?.(payload.data as Subtitle);
          break;
        case "subtitle.failed":
          await onSubtitleFailed?.(
            payload.data as { job_id: string; error: string },
          );
          break;
        case "batch.completed":
          await onBatchCompleted?.(payload.data as BatchExtractResponse);
          break;
      }

      // Call generic handler
      await onAny?.(payload);

      return new Response(JSON.stringify({ received: true }), {
        status: 200,
        headers: { "Content-Type": "application/json" },
      });
    } catch (error) {
      logger.error?.("Webhook handler error", error);
      return new Response(
        JSON.stringify({ error: "Webhook processing failed" }),
        {
          status: 500,
          headers: { "Content-Type": "application/json" },
        },
      );
    }
  };
}

/**
 * Create a webhook handler for Cloudflare Workers
 *
 * @example
 * ```ts
 * import { createCloudflareWebhookHandler } from '@youtube-subtitle-api/sdk/webhook';
 *
 * export default {
 *   async fetch(request: Request, env: { WEBHOOK_SECRET: string }) {
 *     if (request.url.endsWith('/webhook')) {
 *       const handler = createCloudflareWebhookHandler({
 *         secret: env.WEBHOOK_SECRET,
 *         onSubtitleCompleted: (result) => {
 *           console.log('Subtitles extracted:', result.video_id);
 *         },
 *       });
 *       return handler(request);
 *     }
 *     return new Response('Not found', { status: 404 });
 *   },
 * };
 * ```
 */
export function createCloudflareWebhookHandler(
  config: WebhookHandlerConfig,
): (request: Request) => Promise<Response> {
  const handler = createWebhookHandler(config);

  return async (request: Request): Promise<Response> => {
    const body = await request.arrayBuffer();
    const headers = Object.fromEntries(request.headers.entries());

    return handler({
      body: Buffer.from(body),
      headers,
    });
  };
}
